editor/src/components/context-menu-wrapper.tsx

```diff
@@ -18,206 +27,195 @@ import type { WindowPoint } from '../core/shared/math-utils'
         } else {
-          splitItems.push({
+          tempItems.push({
             type: 'submenu',
-            label: item.submenuName!,
+            label: item.submenuName,
             items: [item],
           })
         }
       } else {
-        splitItems.push({
+        tempItems.push({
           type: 'simple',
           item: item,
         })
       }
     }
-    return splitItems
-  }
-
-  isHidden = (item: ContextMenuItem<T>): (() => boolean) => {
-    return () => {
-      if (item.isHidden == null) {
-        return false
-      } else if (typeof item.isHidden === 'function') {
-        return item.isHidden(this.props.getData())
-      } else {
-        return item.isHidden
-      }
-    }
-  }
+    return tempItems
+  }, [items])
 
-  isDisabled = (item: ContextMenuItem<T>): (() => boolean) => {
-    return () => {
-      if (typeof item.enabled === 'function') {
-        return !item.enabled(this.props.getData())
-      } else {
-        return !item.enabled
+  const isHidden = useCallback(
+    (item: Item<T>): (() => boolean) => {
+      return () => {
+        if (typeof item?.isHidden === 'function') {
+          return item.isHidden(getData())
+        }
+        return item?.isHidden ?? false
       }
-    }
-  }
+    },
+    [getData],
+  )
 
-  renderItem(item: ContextMenuItem<T>, index: number) {
-    return (
-      <Item
-        key={`context-menu-${index}-item`}
-        disabled={this.isDisabled(item)}
-        // eslint-disable-next-line react/jsx-no-bind
-        onClick={({ event, triggerEvent }) => {
-          event.stopPropagation()
-          const rightClickCoordinate: WindowPoint | null = (() => {
-            if (!(triggerEvent instanceof MouseEvent)) {
-              return null
-            }
-            return windowPoint({ x: triggerEvent.clientX, y: triggerEvent.clientY })
-          })()
-          item.action(this.props.getData(), this.props.dispatch, rightClickCoordinate, event)
-          contextMenu.hideAll()
-        }}
-        hidden={this.isHidden(item)}
-        style={{
-          height: item.isSeparator ? 9 : 28,
-          display: 'flex',
-          alignItems: 'center',
-          borderRadius: 4,
-        }}
-      >
-        <span style={{ flexGrow: 1, flexShrink: 0 }} className='react-contexify-span'>
-          {item.name}
-        </span>
-        <span style={{ flexGrow: 0, flexShrink: 0, opacity: 0.6 }} className='shortcut'>
-          {item.shortcut}
-        </span>
-      </Item>
-    )
-  }
+  const isDisabled = useCallback(
+    (item: Item<T>): (() => boolean) =>
+      () => {
+        return typeof item?.enabled === 'function' ? !item.enabled(getData?.()) : !item?.enabled
+      },
+    [getData],
+  )
 
-  onShown = () => {
-    document.body.classList.add(BodyMenuOpenClass)
-  }
-  onHidden = () => {
-    document.body.classList.remove(BodyMenuOpenClass)
-  }
+  const renderItem = useCallback(
+    (item: Item<T>, index: number) => {
+      return (
```

Comment:
how easy/hard will it be to extract it to its own component instead of a callback?

Reply:
I probably should do this.. I'll update!

Reply:
So... this felt easy and like a no brainer for organization. But I feel like there are just hidden traps waiting for me in this component 😂. I spent all afternoon trying to get it and making it a new component broke the 'Paste' functionality and tests. So I've reverted that.

If you think it's important, maybe we can follow up later.